From 4c43fb8345c6264595af9aba494c15ff262b5493 Mon Sep 17 00:00:00 2001
From: David Keogh <davidtkeogh@gmail.com>
Date: Sun, 10 Jan 2021 21:26:11 -0800
Subject: appearance: Add option to sync xfwm4 theme

Closes #164, Merge request !43

If the "xfwm4" binary is found in path an option is shown to also set
the xfwm4 theme of the same name when a Gtk theme is set.
In addition the Gtk theme list now contains more information on what
types of support a certain theme contains (Gtk3, Gtk2, Xfwm4,
Xfce4-notifyd).
---
 .../appearance-settings/appearance-dialog.glade    | 54 +++++++++++---
 dialogs/appearance-settings/main.c                 | 82 +++++++++++++++++++---
 2 files changed, 117 insertions(+), 19 deletions(-)

diff --git a/dialogs/appearance-settings/appearance-dialog.glade b/dialogs/appearance-settings/appearance-dialog.glade
index 59234a0b..6e924f7c 100644
--- a/dialogs/appearance-settings/appearance-dialog.glade
+++ b/dialogs/appearance-settings/appearance-dialog.glade
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- Generated with glade 3.22.1 -->
+<!-- Generated with glade 3.22.2 -->
 <interface>
   <requires lib="gtk+" version="3.20"/>
   <requires lib="libxfce4ui-2" version="4.13"/>
@@ -122,34 +122,34 @@
             <property name="can_focus">False</property>
             <property name="layout_style">end</property>
             <child>
-              <object class="GtkButton" id="button1">
-                <property name="label" translatable="yes">_Close</property>
+              <object class="GtkButton" id="button2">
+                <property name="label" translatable="yes">_Help</property>
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
                 <property name="receives_default">True</property>
-                <property name="image">image1</property>
+                <property name="image">image2</property>
                 <property name="use_underline">True</property>
               </object>
               <packing>
                 <property name="expand">False</property>
                 <property name="fill">False</property>
                 <property name="position">0</property>
+                <property name="secondary">True</property>
               </packing>
             </child>
             <child>
-              <object class="GtkButton" id="button2">
-                <property name="label" translatable="yes">_Help</property>
+              <object class="GtkButton" id="button1">
+                <property name="label" translatable="yes">_Close</property>
                 <property name="visible">True</property>
                 <property name="can_focus">True</property>
                 <property name="receives_default">True</property>
-                <property name="image">image2</property>
+                <property name="image">image1</property>
                 <property name="use_underline">True</property>
               </object>
               <packing>
                 <property name="expand">False</property>
                 <property name="fill">False</property>
                 <property name="position">0</property>
-                <property name="secondary">True</property>
               </packing>
             </child>
           </object>
@@ -232,6 +232,42 @@
                     <property name="position">1</property>
                   </packing>
                 </child>
+                <child>
+                  <object class="GtkBox" id="xfwm4_sync">
+                    <property name="visible">True</property>
+                    <property name="can_focus">False</property>
+                    <property name="margin_top">12</property>
+                    <child>
+                      <object class="GtkLabel">
+                        <property name="visible">True</property>
+                        <property name="can_focus">False</property>
+                        <property name="label" translatable="yes">Set matching Xfwm4 theme if there is one</property>
+                        <property name="xalign">0</property>
+                      </object>
+                      <packing>
+                        <property name="expand">True</property>
+                        <property name="fill">True</property>
+                        <property name="position">0</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkSwitch" id="xfwm4_sync_switch">
+                        <property name="visible">True</property>
+                        <property name="can_focus">True</property>
+                      </object>
+                      <packing>
+                        <property name="expand">False</property>
+                        <property name="fill">True</property>
+                        <property name="position">1</property>
+                      </packing>
+                    </child>
+                  </object>
+                  <packing>
+                    <property name="expand">False</property>
+                    <property name="fill">True</property>
+                    <property name="position">2</property>
+                  </packing>
+                </child>
               </object>
             </child>
             <child type="tab">
@@ -921,8 +957,8 @@
       </object>
     </child>
     <action-widgets>
-      <action-widget response="0">button1</action-widget>
       <action-widget response="-11">button2</action-widget>
+      <action-widget response="0">button1</action-widget>
     </action-widgets>
   </object>
 </interface>
diff --git a/dialogs/appearance-settings/main.c b/dialogs/appearance-settings/main.c
index 39c867a7..c3d25c05 100644
--- a/dialogs/appearance-settings/main.c
+++ b/dialogs/appearance-settings/main.c
@@ -64,7 +64,7 @@ enum
     COLUMN_THEME_NAME,
     COLUMN_THEME_DISPLAY_NAME,
     COLUMN_THEME_COMMENT,
-    COLUMN_THEME_NO_CACHE,
+    COLUMN_THEME_WARNING,
     N_THEME_COLUMNS
 };
 
@@ -185,6 +185,7 @@ cb_theme_tree_selection_changed (GtkTreeSelection *selection,
 {
     GtkTreeModel *model;
     gboolean      has_selection;
+    gboolean      has_xfwm4;
     gchar        *name;
     GtkTreeIter   iter;
 
@@ -192,12 +193,27 @@ cb_theme_tree_selection_changed (GtkTreeSelection *selection,
     has_selection = gtk_tree_selection_get_selected (selection, &model, &iter);
     if (G_LIKELY (has_selection))
     {
-        /* Get the theme name */
-        gtk_tree_model_get (model, &iter, COLUMN_THEME_NAME, &name, -1);
+        has_xfwm4 = FALSE;
+
+        /* Get the theme name and whether there is a xfwm4 theme as well */
+        gtk_tree_model_get (model, &iter, COLUMN_THEME_NAME, &name, COLUMN_THEME_WARNING, &has_xfwm4, -1);
 
         /* Store the new theme */
         xfconf_channel_set_string (xsettings_channel, property, name);
 
+        /* Set the matching xfwm4 theme if the selected theme: is not an icon theme,
+         * the xfconf setting is on, and a matching theme is available */
+        if (xfconf_channel_get_bool (xsettings_channel, "/Net/SyncThemes", TRUE) == TRUE
+            && strcmp (property, "/Net/ThemeName") == 0)
+        {
+            if (!has_xfwm4)
+                xfconf_channel_set_string (xfconf_channel_get ("xfwm4"), "/general/theme", name);
+
+            /* Use the default theme if Adwaita is selected */
+            else if (strcmp (name, "Adwaita") == 0 || strcmp(name, "Adwaita-dark") == 0)
+                xfconf_channel_set_string (xfconf_channel_get ("xfwm4"), "/general/theme", "Default");
+        }
+
         /* Cleanup */
         g_free (name);
     }
@@ -460,7 +476,7 @@ appearance_settings_load_icon_themes (gpointer user_data)
                                         COLUMN_THEME_PREVIEW, preview,
                                         COLUMN_THEME_NAME, file,
                                         COLUMN_THEME_DISPLAY_NAME, visible_name,
-                                        COLUMN_THEME_NO_CACHE, !has_cache,
+                                        COLUMN_THEME_WARNING, !has_cache,
                                         COLUMN_THEME_COMMENT, cache_tooltip,
                                         -1);
 
@@ -524,9 +540,15 @@ appearance_settings_load_ui_themes (gpointer user_data)
     gchar        *active_theme_name;
     gchar        *gtkrc_filename;
     gchar        *gtkcss_filename;
+    gchar        *xfwm4_filename;
+    gchar        *notifyd_filename;
+    gchar        *theme_name_markup;
     gchar        *comment_escaped;
     gint          i;
     GSList       *check_list = NULL;
+    gboolean      has_xfwm4;
+    gboolean      has_gtk2;
+    gboolean      has_notifyd;
 
     list_store = pd->list_store;
     tree_view = pd->tree_view;
@@ -552,11 +574,13 @@ appearance_settings_load_ui_themes (gpointer user_data)
         /* Iterate over filenames in the directory */
         while ((file = g_dir_read_name (dir)) != NULL)
         {
-            /* Build the theme style filename */
+            /* Build the filenames for theme components */
             gtkrc_filename = g_build_filename (ui_theme_dirs[i], file, "gtk-2.0", "gtkrc", NULL);
             gtkcss_filename = g_build_filename (ui_theme_dirs[i], file, "gtk-3.0", "gtk.css", NULL);
+            xfwm4_filename = g_build_filename (ui_theme_dirs[i], file, "xfwm4", "themerc", NULL);
+            notifyd_filename = g_build_filename (ui_theme_dirs[i], file, "xfce-notify-4.0", "gtk.css", NULL);
 
-            /* Check if the gtkrc file exists and the theme is not already in the list */
+            /* Check if the gtk.css file exists and the theme is not already in the list */
             if (g_file_test (gtkcss_filename, G_FILE_TEST_EXISTS)
                 && g_slist_find_custom (check_list, file, (GCompareFunc) g_utf8_collate) == NULL)
             {
@@ -585,17 +609,41 @@ appearance_settings_load_ui_themes (gpointer user_data)
                     comment_escaped = NULL;
                 }
 
+                /* Check if the xfwm4 themerc, gtk2 gtkrc, etc. files exist */
+                has_xfwm4 = FALSE;
+                has_gtk2 = FALSE;
+                has_notifyd = FALSE;
+
+                if (g_file_test (xfwm4_filename, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_REGULAR))
+                    has_xfwm4 = TRUE;
+                if (g_file_test (gtkrc_filename, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_REGULAR))
+                    has_gtk2 = TRUE;
+                if (g_file_test (notifyd_filename, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_REGULAR))
+                    has_notifyd = TRUE;
+
+                /* Compose the final markup text */
+                theme_name_markup = g_strdup_printf ("<b>%s</b>\nGtk3", theme_name);
+
+                if (has_gtk2)
+                    theme_name_markup = g_strconcat (theme_name_markup, ", Gtk2", NULL);
+                if (has_xfwm4)
+                    theme_name_markup = g_strconcat (theme_name_markup, ", Xfwm4", NULL);
+                if (has_notifyd)
+                    theme_name_markup = g_strconcat (theme_name_markup, ", Xfce4-notifyd", NULL);
+
                 /* Append ui theme to the list store */
                 gtk_list_store_append (list_store, &iter);
                 gtk_list_store_set (list_store, &iter,
                                     COLUMN_THEME_NAME, file,
-                                    COLUMN_THEME_DISPLAY_NAME, theme_name,
+                                    COLUMN_THEME_DISPLAY_NAME, theme_name_markup,
+                                    COLUMN_THEME_WARNING, !has_xfwm4,
                                     COLUMN_THEME_COMMENT, comment_escaped, -1);
 
                 /* Cleanup */
                 if (G_LIKELY (index_file != NULL))
                     xfce_rc_close (index_file);
                 g_free (comment_escaped);
+                g_free (theme_name_markup);
 
                 /* Check if this is the active theme, if so, select it */
                 if (G_UNLIKELY (g_utf8_collate (file, active_theme_name) == 0))
@@ -610,9 +658,11 @@ appearance_settings_load_ui_themes (gpointer user_data)
                 g_free (index_filename);
             }
 
-            /* Free gtkrc filename */
+            /* Free filenames */
             g_free (gtkrc_filename);
             g_free (gtkcss_filename);
+            g_free (xfwm4_filename);
+            g_free (notifyd_filename);
         }
 
         /* Close directory handle */
@@ -1047,7 +1097,7 @@ appearance_settings_dialog_configure_widgets (GtkBuilder *builder)
     /* Warning Icon */
     renderer = gtk_cell_renderer_pixbuf_new ();
     gtk_tree_view_column_pack_start (column, renderer, FALSE);
-    gtk_tree_view_column_set_attributes (column, renderer, "visible", COLUMN_THEME_NO_CACHE, NULL);
+    gtk_tree_view_column_set_attributes (column, renderer, "visible", COLUMN_THEME_WARNING, NULL);
     g_object_set (G_OBJECT (renderer), "icon-name", "dialog-warning", NULL);
 
     pd = preview_data_new (GTK_LIST_STORE (list_store), GTK_TREE_VIEW (object));
@@ -1088,7 +1138,7 @@ appearance_settings_dialog_configure_widgets (GtkBuilder *builder)
     /* Theme Name and Description */
     renderer = gtk_cell_renderer_text_new ();
     gtk_tree_view_column_pack_start (column, renderer, TRUE);
-    gtk_tree_view_column_set_attributes (column, renderer, "text", COLUMN_THEME_DISPLAY_NAME, NULL);
+    gtk_tree_view_column_set_attributes (column, renderer, "markup", COLUMN_THEME_DISPLAY_NAME, NULL);
     g_object_set (G_OBJECT (renderer), "ellipsize", PANGO_ELLIPSIZE_END, NULL);
 
     pd = preview_data_new (list_store, GTK_TREE_VIEW (object));
@@ -1112,6 +1162,18 @@ appearance_settings_dialog_configure_widgets (GtkBuilder *builder)
     g_object_set (object, "name", "Gtk", NULL);
     g_signal_connect (G_OBJECT (object), "clicked", G_CALLBACK (appearance_settings_install_theme_cb), builder);
 
+    /* Switch for xfwm4 theme matching, gets hidden if xfwm4 is not installed */
+    if (g_find_program_in_path ("xfwm4"))
+    {
+        object = gtk_builder_get_object (builder, "xfwm4_sync_switch");
+        xfconf_g_property_bind (xsettings_channel, "/Net/SyncThemes", G_TYPE_BOOLEAN, G_OBJECT (object), "state");
+    }
+    else
+    {
+        object = gtk_builder_get_object (builder, "xfwm4_sync");
+        gtk_widget_hide (GTK_WIDGET (object));
+    }
+
     /* Subpixel (rgba) hinting Combo */
     object = gtk_builder_get_object (builder, "xft_rgba_store");
 
-- 
cgit v1.2.3

